test/unit/test_auth_webbrowser.py

```diff
@@ -34,110 +35,409 @@
+
+    return recv_side_effect
+
+def recv_setup_with_msg_nowait(ref_token, number_of_blocking_io_errors_before_success = 1):
+    call_number = 0
+
+    def internally_scoped_function(*args):
+        nonlocal call_number
+        call_number += 1
+
+        # if we should NOT block, then the MSG_DONTWAIT flag should be second arg
+        if (len(args) > 1 and args[1] == socket.MSG_DONTWAIT):
+            if call_number <= number_of_blocking_io_errors_before_success:
+                raise BlockingIOError()
+            else:
+                return successful_web_callback(ref_token)
+        else:
+            raise Exception(f'socket_client.recv call expected the second arg to be socket.MSG_DONTWAINT, but received: {args}')
+
+    return internally_scoped_function
 
 def test_auth_webbrowser_get():
     """Authentication by WebBrowser positive test case."""
     ref_token = "MOCK_TOKEN"
+    rest = _init_rest(REF_SSO_URL, REF_PROOF_KEY)
+
+    # mock socket
+    mock_socket_pkg = _init_socket(recv_side_effect_func = recv_setup([ successful_web_callback(ref_token) ]))
+
+    # mock webbrowser
+    mock_webbrowser = MagicMock()
+    mock_webbrowser.open_new.return_value = True
 
+    # Mock select.select to return socket client
+    with mock.patch('select.select', return_value=([mock_socket_pkg.return_value], [], [])):
+        auth = AuthByWebBrowser(
+            application=APPLICATION,
+            webbrowser_pkg=mock_webbrowser,
+            socket_pkg=mock_socket_pkg,
+        )
+        auth.prepare(
+            conn=rest._connection,
+            authenticator=AUTHENTICATOR,
+            service_name=SERVICE_NAME,
+            account=ACCOUNT,
+            user=USER,
+            password=PASSWORD,
+        )
+        assert not rest._connection.errorhandler.called  # no error
+        assert auth.assertion_content == ref_token
+        body = {"data": {}}
+        auth.update_body(body)
+        assert body["data"]["TOKEN"] == ref_token
+        assert body["data"]["PROOF_KEY"] == REF_PROOF_KEY
+        assert body["data"]["AUTHENTICATOR"] == EXTERNAL_BROWSER_AUTHENTICATOR
+
+def test_auth_webbrowser_socket_recv_retries_up_to_15_times_on_empty_bytearray():
+    """Authentication by WebBrowser retries on empty bytearray response from socket.recv"""
+    ref_token = "MOCK_TOKEN"
     rest = _init_rest(REF_SSO_URL, REF_PROOF_KEY)
 
+    # mock socket
+    mock_socket_pkg = _init_socket(recv_side_effect_func = recv_setup(
+        # 14th return is empty byte array, but 15th call will return successful_web_callback
+        ([bytearray()] * 14) + [ successful_web_callback(ref_token) ]
+    ))
+
     # mock webbrowser
     mock_webbrowser = MagicMock()
     mock_webbrowser.open_new.return_value = True
 
+    # Mock select.select to return socket client
+    with mock.patch('select.select', return_value=([mock_socket_pkg.return_value], [], [])), \
+         mock.patch('time.sleep' ) as sleep:
```

Comment:
I never knew that `mock.patch` had a context manager API. That'll make life easier :)

Reply:
Same.  I stumbled on that as I was searching for a good pattern.  I guess one benefit of not knowing the libraries and frameworks well is that it forced me to dig into the docs a bit to figure out how to get the job done.

Glad it will help!